engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/TypeOfNode.java

```diff
@@ -0,0 +1,44 @@
+package org.enso.interpreter.node.expression.builtin.meta;
+
+import com.oracle.truffle.api.CompilerDirectives;
+import com.oracle.truffle.api.interop.InteropLibrary;
+import com.oracle.truffle.api.interop.UnsupportedMessageException;
+import com.oracle.truffle.api.nodes.Node;
+import com.oracle.truffle.api.profiles.BranchProfile;
+import org.enso.interpreter.Constants;
+import org.enso.interpreter.dsl.BuiltinMethod;
+import org.enso.interpreter.runtime.Context;
+import org.enso.interpreter.runtime.builtin.Builtins;
+import org.enso.interpreter.runtime.error.PanicException;
+import org.enso.interpreter.runtime.library.dispatch.TypesLibrary;
+
+@BuiltinMethod(
+    type = "Meta",
+    name = "type_of_builtin",
+    description = "Returns the type of a value.")
+public class TypeOfNode extends Node {
+  private @Child InteropLibrary library =
+      InteropLibrary.getFactory().createDispatched(Constants.CacheSizes.BUILTIN_INTEROP_DISPATCH);
+  private @Child TypesLibrary types =
+      TypesLibrary.getFactory().createDispatched(Constants.CacheSizes.BUILTIN_INTEROP_DISPATCH);
+
+  Object execute(Object value) {
+    if (library.hasMetaObject(value)) {
+      try {
+        return library.getMetaObject(value);
+      } catch (UnsupportedMessageException e) {
+        CompilerDirectives.transferToInterpreter();
+        Builtins builtins = Context.get(this).getBuiltins();
+        throw new PanicException(
+            builtins.error().makeTypeError(builtins.any(), value, "object"), this);
```

Comment:
Do I understand correctly that this should basically never happen?

In that case - why is that a type error? Shouldn't it be an error with a message that says that something impossible happened so that the user knows it's most likely an internal bug and not their fault?